MigrateServicePreferenceFromLegacy/handler.ts

```diff
@@ -0,0 +1,136 @@
+
+const COSMOS_ERROR_KIND = "COSMOS_ERROR_RESPONSE";
+const CONFLICT_CODE = 409;
+const LOG_PREFIX = "MigrateServicePreferenceFromLegacy";
+
+export const MigrateServicesPreferencesQueueMessage = t.interface({
+  newProfile: RetrievedProfile,
+  oldProfile: RetrievedProfile
+});
+export type MigrateServicesPreferencesQueueMessage = t.TypeOf<
+  typeof MigrateServicesPreferencesQueueMessage
+>;
+
+function isCosmosError(
+  ce: CosmosErrors
+): ce is ReturnType<typeof CosmosErrorResponse> {
+  return ce.kind === COSMOS_ERROR_KIND;
+}
+
+export const createServicePreference = (
+  serviceId: ServiceId,
+  blockedChannels: ReadonlyArray<BlockedInboxOrChannelEnum>,
+  fiscalCode: FiscalCode,
+  version: NonNegativeInteger
+): NewServicePreference => ({
+  fiscalCode,
+  id: makeServicesPreferencesDocumentId(fiscalCode, serviceId, version),
+  isEmailEnabled: !blockedChannels.some(
+    channel => channel === BlockedInboxOrChannelEnum.EMAIL
+  ),
+  isInboxEnabled: !blockedChannels.some(
+    channel => channel === BlockedInboxOrChannelEnum.INBOX
+  ),
+  isWebhookEnabled: !blockedChannels.some(
+    channel => channel === BlockedInboxOrChannelEnum.WEBHOOK
+  ),
+  kind: "INewServicePreference",
+  serviceId,
+  settingsVersion: version
+});
+
+export const blockedsToServicesPreferences = (
+  blocked: {
+    [x: string]: readonly BlockedInboxOrChannelEnum[];
+  },
+  fiscalCode: FiscalCode,
+  version: NonNegativeInteger
+) =>
+  o
+    .fromNullable(blocked)
+    .map(b =>
+      Object.entries(b)
+        .filter(([serviceId, _]) => ServiceId.is(serviceId))
+        .map(
+          ([serviceId, blockedInboxOrChannelsForService]) =>
+            createServicePreference(
+              serviceId as ServiceId,
+              blockedInboxOrChannelsForService,
+              fiscalCode,
+              version
+            ) // cast required: ts do not identify filter as a guard
+        )
+    )
+    .getOrElse([]);
+
+export const MigrateServicePreferenceFromLegacy = (
+  servicePreferenceModel: ServicesPreferencesModel
+) => async (context: Context, input: unknown) =>
+  te
+    .fromEither(MigrateServicesPreferencesQueueMessage.decode(input))
+    .mapLeft(
+      ResponseErrorFromValidationErrors(MigrateServicesPreferencesQueueMessage)
+    )
+    .mapLeft(err => new Error(err.detail))
+    .filterOrElse(
+      migrateInput =>
+        NonNegativeInteger.is(
+          migrateInput.newProfile.servicePreferencesSettings.version
+        ),
+      new Error("Can not migrate to negative services preferences version.")
+    )
+    .map(migrateInput =>
+      blockedsToServicesPreferences(
+        migrateInput.oldProfile.blockedInboxOrChannels,
+        migrateInput.newProfile.fiscalCode,
+        /* tslint:disable-next-line no-useless-cast */
+        migrateInput.newProfile.servicePreferencesSettings
+          .version as NonNegativeInteger // cast required: ts do not identify filterOrElse as a guard
+      )
+    )
```

Comment:
we can benefit type narrowing by using an `if` in the same scope:
```suggestion
    .chain(migrateInput =>
      NonNegativeInteger.is(
        migrateInput.newProfile.servicePreferencesSettings.version
      )
        ? te.taskEither.of(
            blockedsToServicesPreferences(
              migrateInput.oldProfile.blockedInboxOrChannels,
              migrateInput.newProfile.fiscalCode,
              migrateInput.newProfile.servicePreferencesSettings.version
            )
          )
        : te.fromLeft<Error, ReturnType<typeof blockedsToServicesPreferences>>(
            new Error(
              "Can not migrate to negative services preferences version."
            )
          )
    )
```

Reply:
There is a functional vs imperative match: is better using a safe cast or stepping away the functional paradigm?